### DISCUSSION TITLE
How to refresh the texture for DirectX12?

### DISCUSSION THREAD

**Discussion by KADGang** [ocornut/imgui#8096] [New Users Build/Link/Run issues ONLY!]
I'm working on a feature similar to an image viewer: the user enters the path to an image and I render it on the screen. I looked at the sample code for this. I plugged it into my source file and it worked perfectly. But when I tried to load another image, it crashed. What should I do in this situation?

My code:

```cpp
{
                    ImGui::Text("CPU handle = %p", my_texture_srv_cpu_handle.ptr);
                    ImGui::Text("GPU handle = %p", my_texture_srv_gpu_handle.ptr);
                    ImGui::Text("size = %d x %d", my_image_width, my_image_height);

                    if (ImGui::Button("load9")) {
                        if (my_texture != NULL) {
                            DestroyTexture(&my_texture);
                            my_image_width = 0;
                            my_image_height = 0;
                        }
                        LoadTextureFromFile("9.png", g_pd3dDevice, my_texture_srv_cpu_handle, &my_texture, &my_image_width, &my_image_height);
                    }

                    ImGui::SameLine();

                    if (ImGui::Button("load8")) {
                        if (my_texture != NULL) {
                            DestroyTexture(&my_texture);
                            my_image_width = 0;
                            my_image_height = 0;
                        }
                        LoadTextureFromFile("chara_ram.png", g_pd3dDevice, my_texture_srv_cpu_handle, &my_texture, &my_image_width, &my_image_height);
                    }

                    ImGui::SameLine();

                    if (ImGui::Button("clear")) {
                        if (my_texture != NULL) {
                            my_texture_srv_gpu_handle = g_pd3dSrvDescHeap->GetGPUDescriptorHandleForHeapStart();
                            my_texture_srv_gpu_handle.ptr += (handle_increment * descriptor_index);

//                            DestroyTexture(&my_texture);
                            my_image_width = 0;
                            my_image_height = 0;
                        }
                    }

                    // Note that we pass the GPU SRV handle here, *not* the CPU handle. We're passing the internal pointer value, cast to an ImTextureID
                    ImGui::Image((ImTextureID)my_texture_srv_gpu_handle.ptr, ImVec2((float)my_image_width, (float)my_image_height));
                }
```

I set three buttons in the window, for the first one and the second one, if you click it then it will load a image file for a texture.

![image](https://github.com/user-attachments/assets/7ddc6feb-4ce3-4c06-9a8d-03862aa089c1)

It works very well at first time to load a image:

![image](https://github.com/user-attachments/assets/3b5af762-20b6-4bb5-b6bb-15a388ae1200)

But when I click the button twice, it crashed. 

![image](https://github.com/user-attachments/assets/c2716bd1-4581-452f-9cc2-d643d1e13d11)

Could you tell me how to implement this function correctly?

**Comment by KADGang** [comment]
I finally done it, I moved the code into my inner source file then it worked. Unbelievable 🤣

**Reply by PathogenDavid** [reply]
It's likely only working as a coincidence.

At a glance it looks like you're destroying the texture without regarding whether it's still in use by the GPU. You either need to block the CPU until the GPU is done with the texture (IE: using a fence) or ideally you need more sophisticated resource management using multiple textures instead of just the one.

Your code also tries to draw an image using an invalid descriptor when you click the clear button.

This is more of a Direct3D 12 problem than a Dear ImGui one. You either need to study Direct3D 12 more or switch to Direct3D 11 or OpenGL--both of which handle this sort of thing automatically.

**Reply by KADGang** [reply]
Thanks for your suggestion. I really don't know much about directx12, I will continue to learn it.
